src/ray/gcs/gcs_server/gcs_kv_manager.h

```diff
@@ -14,43 +14,136 @@
 
 #pragma once
 #include <memory>
+
+#include "absl/container/btree_map.h"
+#include "absl/synchronization/mutex.h"
 #include "ray/gcs/redis_client.h"
 #include "ray/gcs/store_client/redis_store_client.h"
 #include "ray/rpc/gcs_server/gcs_rpc_server.h"
 
 namespace ray {
 namespace gcs {
 
+/// \class InternalKVInterface
+/// The interface for internal kv implementation. Ideally we should merge this
+/// with store client, but due to compatibility issue, we keep them separated
+/// right now.
+class InternalKVInterface {
+ public:
+  /// Get the value associated with `key`
+  ///
+  /// \param key The key to fetch
+  /// \param cb Callback function.
+  virtual void Get(const std::string &key,
+                   std::function<void(std::optional<std::string>)> cb) = 0;
+
+  /// Associate key with value.
+  ///
+  /// \param key The key for the pair
+  /// \param value The value for the pair
+  /// \param overwrite If the key has existed before this op, nothing will happen
```

Comment:
```suggestion
  /// \param overwrite Whether to overwrite existing values. Otherwise, the update will be ignored.
```

Reply:
Btw, this is a *very* strange option to silently ignore data--- do we ever set overwrite = False?

Reply:
Yes, we do. We use this as a lock before. It provides the semantics: check-and-set.